projects/graphql-client/src/mutation-tracker/mutation-tracker.service.ts

```diff
@@ -0,0 +1,75 @@
+import { Injectable } from '@angular/core';
+
+@Injectable({ providedIn: 'root' })
+export class MutationTrackerService {
+  /**
+   * This tracker service assumes that a mutation affects two types of queries.
```

Comment:
So some (if not all) of this is certainly part of apollo, but I'm not sure how configurable/easy to use that is - did you explore that path (and if so, what were the issues)?

Reply:
I was trying to get this done last day and seems like this service approach is the better solution.

We have 3 types of queries in our system.
1. ID based queries
2. List queries
3. Root queries (have no ID at all. configs)

I built a solution where we contain the extend of changes to the mutation method. Apollo lets us have an update call back executed after a mutation, using this we can clear cache entries for the objects that were mutated. This works for ID based queries and a modified version of it can work for root queries. However, this approach leads to inconsistent data for list queries. After mutating an object, if we delete its entry, the list query will be unaware of the change and will continue to use the cached list. This means the entry we just updated will be missing from the list until an explicit over the network query is made. So, this solution does not really work well unless we deploy our own custom logic for identifying which entry is associated with which queries and so on.

Given that, I feel like this tracker service approach is safer and cleaner. It doesn't modify any data directly. 

We could explore how to push down the markAsConsumed part down to the graphql service.

Thoughts? 

@aaron-steinfeld

Reply:
I think the main goal is to keep any real logic out of the handlers. If mutation handlers have an abstract method/field they need to implement to say what they affect, and query handlers on what they read, then any cache decisions or evictions can happen at the service, as you indicated. Later, if we find a way to push that logic into apollo, it's a local impl change to that lower level service. I agree there's likely no way to void handlers providing those two pieces of data given how some apis are structured.